src/xmipp/libraries/tomo/tomo_extract_subtomograms.cpp

```diff
@@ -255,105 +294,67 @@ void ProgTomoExtractSubtomograms::run()
 			continue;
 
 		if (fixedBoxSize && dsFactorDiff > dsFactorTolerance)
-		{
 			subtomo.initZeros(1, boxSizeExtraction, boxSizeExtraction, boxSizeExtraction);
-		}
 		else
-		{
 			subtomo.initZeros(1, boxsize, boxsize, boxsize);
-		}
 
-		// Contrast inversion
-		for (int k=zinit; k<zlim; k++)
-		{
-			int kk = k - zcoor;
-			for (int i=yinit; i<ylim; i++)
-			{
-				int ii = i-ycoor;
-				for (int j=xinit; j<xlim; j++)
-				{
-					A3D_ELEM(subtomo, kk+halfboxsize, ii+halfboxsize, j+halfboxsize-xcoor) = invertSign * A3D_ELEM(tom, k, i, j);
-				}
-			}
-		}
+		extractSubtomo(tom, subtomo,  xinit, yinit, zinit, invertSign);
 
 		if (fixedBoxSize && dsFactorDiff > dsFactorTolerance)
-		{
 			extractSubtomoFixedSize(subtomo);
-		}
 
 		// Normalization
 		if (normalize)
-		{
 			normalizeSubtomo(subtomo, halfboxsize);
-		}
 
-		size_t particleid = -1;
-		if (row.containsLabel(MDL_PARTICLE_ID))
-		{
-			row.getValue(MDL_PARTICLE_ID, particleid);
-		}
 
-		writeSubtomo(idx, xcoor, ycoor, zcoor, particleid);
+		listOfSubtomos.push_back(subtomo);
 
-		++idx;
-	}
+		#ifdef DEBUG
+		std::cout << fn << std::endl;
+		std::cout << fnOut << std::endl;
+		std::cout << fnOut+fn << std::endl;
+		#endif
 
+		//subtomoImg.write(fnOut+"/"+fn);
 
-	std::cout << "Subtomo substraction finished succesfully!!" << std::endl;
-}
+		FileName composedFn;
+		composedFn.compose(idx, fn);
+		rowout = row;
+		rowout.setValue(MDL_IMAGE, composedFn);
+		mdout.addRow(rowout);
 
+		++idx;
 
-void ProgTomoExtractSubtomograms::upsample(const MultidimArray<std::complex<double>> &from, MultidimArray<std::complex<double>> &to)
-{
-	for (size_t k = 0; k < ZSIZE(from)/2; k++)
-	{
-		for (size_t j = 0; j < YSIZE(from)/2; j++)
-		{
-			for (size_t i = 0; i < XSIZE(from); i++)
-			{
-				// Origin cube
-				DIRECT_A3D_ELEM(to, k, j, i) = DIRECT_A3D_ELEM(from, k, j, i);
+	}
+
+	size_t numberOfSubtomos = listOfSubtomos.size();
 
-				// Oposite cube
-				DIRECT_A3D_ELEM(to,  ZSIZE(to) - ZSIZE(from)/2 + k, YSIZE(to) - YSIZE(from)/2 + j, i) =
-				DIRECT_A3D_ELEM(from, ZSIZE(from)/2 + k, YSIZE(from)/2 + j, i);
+	MultidimArray<float> subtomosStack;
 
-				// Offset-Z cube
-				DIRECT_A3D_ELEM(to,  ZSIZE(to) - ZSIZE(from)/2 + k, j, i) =
-				DIRECT_A3D_ELEM(from, ZSIZE(from)/2 + k, j, i);
+	subtomosStack.initZeros(numberOfSubtomos, boxsize, boxsize, boxsize);
 
-				// Offset-Y cube
-				DIRECT_A3D_ELEM(to,  k, YSIZE(to) - YSIZE(from)/2 + j, i) =
-				DIRECT_A3D_ELEM(from, k, YSIZE(from)/2 + j, i);
+	for (size_t subtomoIdx=0; subtomoIdx<numberOfSubtomos; subtomoIdx++)
```

Comment:
Using float for storage does not solve the core problem of excessive memory usage. Subtomos should be written to the volume stack on streaming using `WRITE_APPEND` mode. This reduces memory complexity from O(N) to O(1)